### PR TITLE
Save annotation to snapshot

### PR DESCRIPTION
**Pull request by mtanda**
Fix part of https://github.com/grafana/grafana/issues/2956.
(Not save templating, only save annotation)

### CODE REVIEW

**Comment by bergquist** [comment]
I like the idea and it works good. But I would like to see some tests for this before its merged. Is there anyway to reduce the amount of if statements? can a default values be used instead?

**Comment by mtanda** [comment]
remove one "if" statement.

Currently, there is no test for time series snapshot feature.
I don't have idea how I can test snapshot feature...
Is there good test example? @bergquist @torkelo

**Comment by torkelo** [comment]
I would like to see the annotation definition names (just the name not the query) still visible in the snapshot, and the snapshot data should not be saved in the dashboard top level, it would be better to store the snapshot data on each annotation definition. 

This requires a slight change the the snapshot clean up code, so it will include enabled annotation definitions (but strip out everything except the name and the snapshot data on it)

**Comment by mtanda** [comment]
Thanks for feedback!
I'll change the current code.

**Comment by mtanda** [comment]
@torkelo I update the code. Please review this.

File: public/app/features/annotations/annotationsSrv.js [hunk 1]
```diff
@@ -38,10 +38,19 @@ define([
...
+          self.receiveAnnotationResults(annotation.snapshotData);
+          return;
+        }
         return datasourceSrv.get(annotation.datasource).then(function(datasource) {
           var query = {range: range, rangeRaw: rangeRaw, annotation: annotation};
           return datasource.annotationQuery(query)
             .then(self.receiveAnnotationResults)
+            .then(function(options) {
+              if (dashboard.snapshot) {
+                annotation.snapshotData = angular.copy(options);
```
Comment (torkelo) [inline public/app/features/annotations/annotationsSrv.js]: very small detail: 
options variable was little confusingly named :)

File: public/app/features/dashboard/shareSnapshotCtrl.js [hunk 2]
```diff
@@ -101,8 +101,14 @@ function (angular, _) {
         panel.links = [];
         panel.datasource = null;
       });
-      // remove annotations
-      dash.annotations.list = [];
+      // remove annotation queries
+      dash.annotations.list = _.map(dash.annotations.list, function(annotation) {
```
Comment (torkelo) [inline public/app/features/dashboard/shareSnapshotCtrl.js]: I think you can filter out only enabled ones here.

File: public/app/features/annotations/annotationsSrv.js (hunk 1 above)
Comment (mtanda) [inline public/app/features/annotations/annotationsSrv.js]: Ah, I missed. change "options" to "results".

File: public/app/features/dashboard/shareSnapshotCtrl.js (hunk 2 above)
Comment (mtanda) [inline public/app/features/dashboard/shareSnapshotCtrl.js]: Yes, I can. It's better to filter out here.

**Comment by mtanda** [comment]
Update code again.

**Pull request merged**